### PR TITLE
Test: E2E tests for new Component Testing features

### LINKED ISSUE #29745: E2E tests for Testing improvements
The following new features are not E2E tested right now:

- [x] Running a focused test on a story
- [x] Running a focused test on a component
- [x] Running a focused test on a group
- [ ] ~Changing test config in sidebar context menu (and then running)~
- [x] Changing test config in testing module (and then running)
- [x] Running with coverage - and opening link to HTML report

### PR DESCRIPTION
**Pull request by JReinhold**
Telescopes on to https://github.com/storybookjs/storybook/pull/29808

Closes #29745

## What I did

- Added E2E tests for Focused Tests, Coverage, and a combination of both
- Modified DOM a11y slightly 
- Modified existing E2E tests slightly
- Added tests for `preview-head.html` functionality. They don't need to be E2E tested, they can be tested by just running Vitest. So I added that to CI too, skipping the tests that fail on purpose.

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [X] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.7 MB | 77.7 MB | 0 B | 0.63 | 0% |
| initSize |  136 MB | 136 MB | -1 B | 0.42 | 0% |
| diffSize |  58.4 MB | 58.4 MB | -1 B | 0.42 | 0% |
| buildSize |  7.24 MB | 7.24 MB | -1 B | 0.23 | 0% |
| buildSbAddonsSize |  1.88 MB | 1.88 MB | 0 B | 0.23 | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.86 MB | 1.86 MB | -1 B | 1.02 | 0% |
| buildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.93 MB | 3.93 MB | -1 B | 0.23 | 0% |
| buildPreviewSize |  3.3 MB | 3.3 MB | 0 B | -0.27 | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  26.8s | 7.2s | -19s -653ms | -0.85 | -272.3% |
| generateTime |  19.7s | 20.7s | 970ms | 0.4 | 4.7% |
| initTime |  14.8s | 13.8s | -1s -59ms | -0.33 | -7.7% |
| buildTime |  8.3s | 8.6s | 264ms | -0.77 | 3.1% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  5.1s | 4.9s | -234ms | -0.3 | -4.7% |
| devManagerResponsive |  3.9s | 3.7s | -178ms | -0.07 | -4.7% |
| devManagerHeaderVisible |  884ms | 723ms | -161ms | 1.18 | -22.3% |
| devManagerIndexVisible |  973ms | 751ms | -222ms | 0.98 | -29.6% |
| devStoryVisibleUncached |  1.9s | 2s | 120ms | **1.52** | 🔺5.8% |
| devStoryVisible |  970ms | 752ms | -218ms | 0.96 | -29% |
| devAutodocsVisible |  679ms | 717ms | 38ms | **2.69** | 🔺5.3% |
| devMDXVisible |  718ms | 625ms | -93ms | 0.95 | -14.9% |
| buildManagerHeaderVisible |  633ms | 677ms | 44ms | 0.87 | 6.5% |
| buildManagerIndexVisible |  753ms | 772ms | 19ms | 0.78 | 2.5% |
| buildStoryVisible |  589ms | 641ms | 52ms | 1.07 | 8.1% |
| buildAutodocsVisible |  485ms | 509ms | 24ms | 0.82 | 4.7% |
| buildMDXVisible |  499ms | 498ms | -1ms | 0.4 | -0.2% |

## Greptile Summary

Here's my concise summary of the PR changes:

Added end-to-end tests and infrastructure for Storybook's component testing features, focusing on Vitest integration and coverage reporting.

- Added E2E tests in `portable-stories-kitchen-sink/react/e2e-tests/component-testing.spec.ts` for focused tests and coverage reporting
- Modified `code/addons/test/src/node/vitest-manager.ts` to handle coverage options and test filtering
- Added coverage reporter in `code/addons/test/src/node/coverage-reporter.ts` with watermark support
- Added accessibility improvements to test UI components in `code/addons/test/src/components/TestProviderRender.tsx`
- Added new test stories and configurations in `portable-stories-kitchen-sink/react/` for validating component testing features

### CODE REVIEW

File: test-storybooks/portable-stories-kitchen-sink/react/package.json [hunk 1]
```diff
@@ -108,7 +109,9 @@
     "@typescript-eslint/eslint-plugin": "^6.21.0",
     "@typescript-eslint/parser": "^6.21.0",
     "@vitejs/plugin-react": "^4.2.1",
-    "@vitest/browser": "^2.1.3",
+    "@vitest/browser": "^2.1.6",
+    "@vitest/coverage-v8": "^2.1.6",
+    "@vitest/ui": "2.1.8",
```
Comment (yannbf) [inline test-storybooks/portable-stories-kitchen-sink/react/package.json:114]: you might not want to have mismatching versions here, this should be either ^2.1.6 or all other vitest deps should be 2.1.8

File: code/lib/test/package.json [hunk 2]
```diff
@@ -49,8 +49,8 @@
     "@testing-library/dom": "10.4.0",
     "@testing-library/jest-dom": "6.5.0",
     "@testing-library/user-event": "14.5.2",
-    "@vitest/expect": "2.0.5",
```
Comment (yannbf) [inline code/lib/test/package.json:52]: why was this change needed for writing E2E tests? This can have impact on users, do you what that impact would be?

File: test-storybooks/portable-stories-kitchen-sink/react/package.json (hunk 1 above)
Comment (JReinhold) [reply test-storybooks/portable-stories-kitchen-sink/react/package.json:114]: right, Vitest added the 2.1.8 automatically here.

File: code/lib/test/package.json (hunk 2 above)
Comment (JReinhold) [reply code/lib/test/package.json:52]: I ran into a problem when running the E2E tests, caused by multiple Vitest versions _somewhere_, can't remember where or what the problem was, but it was fixed by aligning all the versions.

I can't think of how this would affect users negatively, we upgrade non-breaking dependencies all the time and I'm not sure how this is different. It was just an attempt to do a bit of house cleaning while touching that code anyway.

But I don't care much about it, I'll revert.

File: test-storybooks/portable-stories-kitchen-sink/react/e2e-tests/component-testing.spec.ts [hunk 3]
```diff
@@ -240,4 +237,223 @@ test.describe("component testing", () => {
...
+    // Assert - 1 failing test shows as a failed status
+    await expect(sidebarContextMenu.getByText('3 stories with errors')).toBeVisible();
+    await expect(sidebarContextMenu.getByLabel('status: failed')).toHaveCount(1);
+
+    await page.click('body');
+    await expect(page.locator('#storybook-explorer-menu').getByRole('status', { name: 'Test status: success' })).toHaveCount(4);
+    await expect(page.locator('#storybook-explorer-menu').getByRole('status', { name: 'Test status: error' })).toHaveCount(2);
+  });
+
+  test("should run focused tests without coverage, even when enabled", async ({
```
Comment (JReinhold) [inline test-storybooks/portable-stories-kitchen-sink/react/e2e-tests/component-testing.spec.ts:404]: this test is slightly excessive.

It:

1. enables coverage
2. do a focus test
3. assert that coverage is not collected
4. do a full test
5. assert that coverage _is_ collected

I'm not sure if we should keep it or not, you could also argue that it's largely covered by the unit test here:

https://github.com/storybookjs/storybook/pull/29749/files#diff-18fe411ce6b83d11273334c241ee323db41b0c412c66ce04ceef562e5e39a899R172-R201

WDYT @yannbf ?

Comment (yannbf) [reply test-storybooks/portable-stories-kitchen-sink/react/e2e-tests/component-testing.spec.ts:404]: If it's too taxing or flake-prone, I'd say remove it. Seems to be well tested elsewhere, but this E2E test does involve more elements in a real scenario. However I think these elements are tested in the other tests above

File: test-storybooks/portable-stories-kitchen-sink/react/package.json [hunk 4]
```diff
@@ -12,8 +12,9 @@
     "playwright-ct": "playwright test -c playwright-ct.config.ts",
     "playwright-e2e": "playwright test -c playwright-e2e.config.ts",
     "preview": "vite preview",
-    "storybook": "storybook dev -p 6006",
-    "vitest": "echo 'not running'"
+    "storybook": "NODE_OPTIONS=\"--preserve-symlinks --preserve-symlinks-main\" storybook dev -p 6006",
```
Comment (JReinhold) [inline test-storybooks/portable-stories-kitchen-sink/react/package.json:15]: this speeds up the story loading of the kitchen sink, at least on my machine.

File: code/addons/test/src/components/TestProviderRender.tsx [hunk 5]
```diff
@@ -307,14 +307,19 @@ export const TestProviderRender: FC<
...
                 <TestStatusIcon
                   percentage={coverageSummary.percentage}
                   status={coverageSummary.status}
                   aria-label={`status: ${coverageSummary.status}`}
                 />
               }
-              right={`${coverageSummary.percentage}%`}
+              right={
+                <span aria-label={`${coverageSummary.percentage} percent coverage`}>
+                  {coverageSummary.percentage} %
```
Comment (ndelangen) [inline code/addons/test/src/components/TestProviderRender.tsx:320]: @ghengeveld

File: code/addons/test/src/components/TestProviderRender.tsx [hunk 6]
```diff
@@ -307,14 +307,19 @@ export const TestProviderRender: FC<
...
+                </span>
```
Comment (ndelangen) [inline code/addons/test/src/components/TestProviderRender.tsx:321]:
```suggestion
                coverageSummary.percentage ? <span aria-label={`${coverageSummary.percentage} percent coverage`}>
                  {coverageSummary.percentage} %
                </span> : null
```

**Review by ndelangen — APPROVED** [review]

File: .circleci/config.yml [hunk 7]
```diff
@@ -757,6 +757,10 @@ jobs:
           name: Run E2E tests
           command: yarn playwright-e2e
           working_directory: test-storybooks/portable-stories-kitchen-sink/react
+      - run:
+          name: Run Vitest
```
Comment (yannbf) [inline .circleci/config.yml:761]: why do we need this?

**Review by yannbf — APPROVED** [review]

File: .circleci/config.yml (hunk 7 above)
Comment (JReinhold) [reply .circleci/config.yml:761]: ah yeah that's much better, it's to test these scenarios: https://github.com/storybookjs/storybook/pull/29826/files#diff-0550ef2a360c5242f25677cd964d3afce1c0412497a76eded7d03c0838941611R68-R92

Another alternative would be to move those stories into the UI Storybook instead, no strong opinion.
